RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc

```diff
@@ -202,35 +239,215 @@ void CAHitQuadrupletGenerator::hitQuadruplets(const TrackingRegion& region,
                            SeedingLayerSetsHits::LayerIndex outer) {
     return pair.innerLayerIndex() == inner && pair.outerLayerIndex() == outer;
   };
-  fillGraph(layers, g, hitDoublets,
-            [&](const SeedingLayerSetsHits::SeedingLayer& inner,
-                const SeedingLayerSetsHits::SeedingLayer& outer,
-                std::vector<const HitDoublets *>& hitDoublets) {
-      using namespace std::placeholders;
-      auto found = std::find_if(regionLayerPairs.begin(), regionLayerPairs.end(),
-                                std::bind(layerPairEqual, _1, inner.index(), outer.index()));
-      if(found != regionLayerPairs.end()) {
-        hitDoublets.emplace_back(&(found->doublets()));
-        return true;
-      }
-      return false;
-    });
 
-  hitQuadruplets(regionLayerPairs.region(), result, hitDoublets, g, es);
+ int index =0;
+ for(const auto& regionLayerPairs: regionDoublets) {
+
+	 const TrackingRegion& region = regionLayerPairs.region();
+	 hitDoublets.clear();
+
+	  if (index == 0){   
+	  	createGraphStructure(layers, g);
+	  }
+	  else{  
+		clearGraphStructure(layers, g);
+	  }
+
+	  fillGraph(layers, g, hitDoublets,
+	            [&](const SeedingLayerSetsHits::SeedingLayer& inner,
+	                const SeedingLayerSetsHits::SeedingLayer& outer,
+	                std::vector<const HitDoublets *>& hitDoublets) {
+	      using namespace std::placeholders;
+	      auto found = std::find_if(regionLayerPairs.begin(), regionLayerPairs.end(),
+	                                std::bind(layerPairEqual, _1, inner.index(), outer.index()));
+	      if(found != regionLayerPairs.end()) {
+	        hitDoublets.emplace_back(&(found->doublets()));
+	        return true;
+	      }
+	      return false;
+	    });
+	edm::ESHandle<TrackerTopology> tTopoHand;
+	es.get<TrackerTopologyRcd>().get(tTopoHand);
+	const TrackerTopology *tTopo=tTopoHand.product();	
+
+
+	const int numberOfHitsInNtuplet = 4;
+	std::vector<CACell::CAntuplet> foundQuadruplets;
+
+	CellularAutomaton ca(g);
+
+	ca.createAndConnectCells(hitDoublets, region, caThetaCut,
+			caPhiCut, caHardPtCut);
+
+	ca.evolve(numberOfHitsInNtuplet);
+
+	ca.findNtuplets(foundQuadruplets, numberOfHitsInNtuplet);
+
+
+	const QuantityDependsPtEval maxChi2Eval = maxChi2.evaluator(es);
+
+ 	// re-used thoughout, need to be vectors because of RZLine interface
+	std::array<float, 4> bc_r;
+	std::array<float, 4> bc_z;
+  	std::array<float, 4> bc_errZ2;
+  	std::array<GlobalPoint, 4> gps;
+  	std::array<GlobalError, 4> ges;
+  	std::array<bool, 4> barrels;
+ 	unsigned int fourthLayerId = 0;
+  	unsigned int previousfourthLayerId = 0;
+  	int subDetId = 0; 
+  	int previousSubDetId = 0;
+ 	unsigned int sideId = 0; 
+  	unsigned int previousSideId = 0;
+ 	std::array<unsigned int, 2> previousCellIds ={{0,0}};
+  	bool isTheSameTriplet = false;
+	bool isTheSameFourthLayer = false;
+	bool hasAlreadyPushedACandidate = false;
+	float selectedChi2 = std::numeric_limits<float>::max();
```

Comment:
Is this really the innermost possible scope for all these variables? I.e. can any of those be moved to inner blocks?

(e.g. `isTheSameTriplet` appears to be used only inside `if(caOnlyOneLastHitPerLayerFilter)`, so better move the variable definition there too)